lighthouse-logger/index.js

```diff
@@ -118,33 +174,145 @@ class Log {
     return level_ === 'verbose';
   }
 
+  /**
+   * @param {Status} param0
+   * @param {LogAction} level
+   */
   static time({msg, id, args = []}, level = 'log') {
     marky.mark(id);
     Log[level]('status', msg, ...args);
   }
 
+  /**
+   * @param {Status} param0
+   * @param {LogAction} level
+   */
   static timeEnd({msg, id, args = []}, level = 'verbose') {
     Log[level]('statusEnd', msg, ...args);
     marky.stop(id);
   }
 
+  /* eslint-disable no-invalid-this */
+  /**
+   * Decorates a function, calling time/timeEnd before/after calling the original function.
+   * @template T, R
+   * @template {*[]} Args
+   * @param {(this: T, ...args: Args) => R} originalFn
+   * @param {TimeDecorateOpts<T, Args>} opts
+   * @return {(this: T, ...args: Args) => R}
+   */
+  static timeDecorate(originalFn, opts) {
+    /**
+     * @type {(this: T, ...args: Args) => R}
+     */
+    const fn = function timeDecoratedFn(...args) {
+      const timeStartLogLevel = opts.timeStartLogLevel || 'log';
+      const timeEndLogLevel = opts.timeEndLogLevel || 'verbose';
+
+      let msg;
+      if (typeof opts.msg === 'string') {
+        msg = opts.msg;
+      } else if (typeof opts.msg === 'function') {
+        msg = opts.msg.apply(this, args);
+      }
+      if (!msg) {
+        throw new Error('expected msg');
+      }
+
+      let id;
+      if (typeof opts.id === 'string') {
+        id = opts.id;
+      } else if (typeof opts.id === 'function') {
+        id = opts.id.apply(this, args);
+      } else {
+        id = `lh:${originalFn.name}`;
+      }
+      if (!id) {
+        throw new Error('expected id');
+      }
+
+      const status = {msg, id};
+      Log.time(status, timeStartLogLevel);
+
+      let result;
+      try {
+        result = originalFn.apply(this, args);
+      } catch (err) {
+        err.stack = err.stack.replace(/.* at timeDecoratedFn .*\n/g, '');
+        throw err;
+      }
+
+      if (result && typeof result.then === 'function') {
+        return result.then((/** @type {any} */ value) => {
+          Log.timeEnd(status, timeEndLogLevel);
+          return value;
+        }).catch((/** @type {any} */ err) => {
+          err.stack = err.stack.replace(/.* at timeDecoratedFn .*\n/, '');
+          throw err;
+        });
+      } else {
+        Log.timeEnd(status, timeEndLogLevel);
+        return result;
+      }
+    };
+    return fn;
+  }
+  /* eslint-enable no-invalid-this */
+
+  /**
+   * Decorates (like timeDecorate) specified methods of a class.
+   * If decorating instance methods, use the class's prototype.
+   * If decorating static methods, use the class directly.
+   * @template {Object|Function} Class
+   * @template {keyof Class} Prop
+   * @param {Class} klass
+   * @param {{[key in Prop]: TimeDecorateOpts<Class, ArgumentTypes<IsFunction<Class[key]>>>}} methods
+   */
+  static timeDecorateClass(klass, methods) {
+    for (const [method, opts] of Object.entries(methods)) {
```

Comment:
~~I'm surprised this works with class prototypes, last time I dealt with this I feel like they weren't enumerable and I had to use `getOwnPropertyNames` for some reason. If you happened to run into some issues that you could share why I struggled last time I'd love to hear your insight :)~~

Ohhhhhh you're iterating over the methods that are manually specified, not the class members! 👍

Reply:
yeah you got it. these are explicitly defined via an object literal when calling the function